tests/common.py

```diff
@@ -31,42 +33,238 @@
+            cls._subdiv_category_lookup,
+            cls._non_public_supported_categories_lookup,
+        ) = cls._get_or_create_lookup(cls.test_class)
+        cls._subdiv_lookup = {**cls._subdiv_base_lookup, **cls._subdiv_category_lookup}
+
+        if years is None:
+            # Default `self.full_range`
+            if not hasattr(cls, "full_range"):
+                # `start_year` and `end_year` are used only if they're included directly in
+                # country/market entities, not inherited from `HolidayBase`.
+                cls.start_year = cls.test_class.__dict__.get("start_year", 1950)
+                cls.end_year = cls.test_class.__dict__.get("end_year", 2049) + 1
+                cls.full_range = range(cls.start_year, cls.end_year)
+            else:
+                cls.start_year = cls.full_range.start
+                cls.end_year = cls.full_range.stop
+            years = cls.full_range
+
+        # Default `years_[insert]` to `years` to prevent redundant initialization.
+        is_observed_subclass = issubclass(cls.test_class, ObservedHolidayBase)
+        is_islamic_subclass = issubclass(cls.test_class, IslamicHolidays)
+        for category in cls._non_public_supported_categories_lookup:
+            year_variants.setdefault(f"years_{category}", years)
+            if is_observed_subclass:
+                year_variants.setdefault(
+                    f"years_{category}_non_observed",
+                    year_variants.get(
+                        "years_non_observed", year_variants.get(f"years_{category}", years)
+                    ),
+                )
+            if is_islamic_subclass:
+                year_variants.setdefault(
+                    f"years_{category}_islamic_no_estimated",
+                    year_variants.get(
+                        "years_islamic_no_estimated", year_variants.get(f"years_{category}", years)
+                    ),
+                )
+
+        # For subdivisions, `years_all_subdivs` can be use for mass-assignments.
+        years_all_subdivs = year_variants.get("years_all_subdivs", years)
+        years_all_subdivs_non_observed = year_variants.get(
+            "years_all_subdivs_non_observed",
+            year_variants.get("years_non_observed", year_variants.get("years_all_subdivs", years)),
+        )
+
+        for key in cls._subdiv_lookup:
+            year_variants.setdefault(f"years_subdiv_{key}", years_all_subdivs)
+            if is_observed_subclass:
+                year_variants.setdefault(
+                    f"years_subdiv_{key}_non_observed", years_all_subdivs_non_observed
+                )
+
+        if is_islamic_subclass:
+            year_variants.setdefault("years_islamic_no_estimated", years)
+        if is_observed_subclass:
+            year_variants.setdefault("years_non_observed", years)
+        if is_islamic_subclass and is_observed_subclass:
+            year_variants.setdefault("years_islamic_no_estimated_non_observed", years)
+
```

Comment:
It would be useful to have the option to skip initialization of certain parameter combinations. For example, in Canada, provinces don't have GOVERNMENT holidays; in most countries, all categories except PUBLIC don't have observed holidays, etc.

Reply:
Maybe something like this?
```python
class TestAustralia(CommonCountryTests, TestCase):
    @classmethod
    def setUpClass(cls):
        super().setUpClass(Australia, with_subdiv_categories=True)
```
We can have the `with_subdiv_categories` as `False` by default, so in this case USA's `setUpClass` stays as-is.
```python
class TestUnitedStates(CommonCountryTests, TestCase):
    @classmethod
    def setUpClass(cls):
        super().setUpClass(UnitedStates, years_non_observed=range(2000, 2024))
```

Alternatively, we can include the categories as well (maybe something like `with_subdiv_categories={BANK, }` for example, but IMO making it a boolean is already suffice for most of our use cases.